analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CloudNative/AzureFunctionsLogFailures.cs

```diff
@@ -0,0 +1,325 @@
+        {
+            void Log() => log.LogError("");
+        }
+    }
+
+    [FunctionName("Sample")]
+    public static void LogExceptionLoggedInLambda(ILogger log)
+    {
+        try { }
+        catch (Exception ex) // FN. Any call to "log" is considered valid. Reachability is not considered.
+        {
+            Action x = () => log.LogError("");
+        }
+    }
+
+    [FunctionName("Sample")]
+    public static void LogExceptionLoggedInAnonymousFunction(ILogger log)
+    {
+        try { }
+        catch (Exception ex) // FN. Any call to "log" is considered valid. Reachability is not considered.
+        {
+            Action x = delegate () { log.LogError(""); };
+        }
+    }
+
+    [FunctionName("Sample")]
+    public static void LogExceptionIsPassedInLambda(ILogger log)
+    {
+        try { }
+        catch (Exception ex) // FN. Any call to "log" is considered valid. Reachability is not considered.
+        {
+            DoSomething(() => log.LogError(""));
+        }
+
+        void DoSomething(Action a) => a();
+    }
+
+    [FunctionName("Sample")]
+    public static void LogExceptionLoggedInUnreachableCode(ILogger log)
+    {
+        try { }
+        catch (Exception ex) // FN. log in unreachable code.
+        {
+            return;
+            log.LogError("");
+        }
+    }
+
+    [FunctionName("Sample")]
+    public static void LogExceptionNotEveryCodePathContainsLog(ILogger log)
+    {
+        try { }
+        catch (Exception ex) // FN. Not every code path contains a log
+        {
+            var i = 100;
+            if (i == 42)
+            {
+                log.LogError("");
+            }
+        }
+    }
+
+    [FunctionName("Sample")]
+    public static void LogExceptionFromWebApi(ILogger log)
+    {
+        try { }
+        catch (Exception ex) // Compliant. Call to unregcognized extension method from WebApi
+        {
+            var i = 100;
+            if (i == 42)
+            {
+                log.LogMetric("Metric", 5);
+            }
+        }
+    }
+
+    [FunctionName("Sample")]
+    public static void LogExceptionInNestedCatch(ILogger log)
+    {
+        try { }
+        catch (Exception ex) // FN. The outer block should contain a log call too
+        {
+            try { }
+            catch
+            {
+                log.LogError("");
+            }
+        }
+    }
+
+    [FunctionName("Sample")]
+    public static void LoggerGetsPassedToFunction(ILogger log)
+    {
+        try { }
+        catch (Exception ex)
+        {
+            LoggerHelper(log); // Compliant. We don't follow the call, but we assume some decent logging takes place, if the logger gets passed along.
+        }
+    }
+
+    [FunctionName("Sample")]
+    public static void WrappedLoggerGetsPassedToFunction(ILogger log)
+    {
+        try { }
+        catch (Exception ex)
+        {
+            LoggerHelper(NullLogger.Instance); // Compliant. Some ILogger is passed
```

Comment:
I would challenge this case. There's an explicit `ILogger log` in the argument. And it is not used at all. I think we should raise an issue.

Reply:
Fair enough. There can be many ILogger in scope, especially with dependency injection. At the moment we check if the type of `argument.Expression` is `ILogger`. We don't look at the symbol of `argument.Expression` and where it comes from (in the case of DI it could be an e.g. field or property, here it is a parameter). The `LoggerIsInScopeInEntryPoint` actually already collects this information and it could be reused here. This would complicate the solution and could let to all kinds of FP (like `NullLogger.Instance` is hard to find as an "ILogger" that is in scope). As long as a logger is passed around, I would bail out.